mongodb-nosqli.yaml

```diff
@@ -0,0 +1,109 @@
+rules:
+- id: mongodb-nosqli
+  message: Detected non-constant data passed into a NoSQL query using the 'where'
+    evaluation operator. If this data can be controlled by an external user, this
+    is a NoSQL injection. Ensure data passed to the NoSQL query is not user controllable,
+    or properly sanitize the data. Ideally, avoid using the 'where' operator at all
+    and instead use the helper methods provided by com.mongodb.client.model.Filters
+    with comparative operators such as eq, ne, lt, gt, etc.
+  languages:
+  - java
+  severity: WARNING
+  patterns:
+  - pattern-either:
+    - pattern: (BasicDBObject $QUERY).put("$where", $INPUT);
+    - pattern: '(HashMap<String, String> $MAP).put("$where", $INPUT);
+
+        ...
+
+        (BasicDBObject $QUERY).putAll($MAP);
+
+        '
+    - pattern: (BasicDBObject $QUERY).append("$where", $INPUT);
+    - pattern: new BasicDBObject("$where", $INPUT);
+    - pattern: '(HashMap<String, String> $MAP).put("$where", $INPUT);
```

Comment:
I might be the playground that reformatted this, but for readability I would reformat these using the `|` operator:
```yaml
    - pattern: |
        (HashMap<String, String> $MAP).put("$where", $INPUT);
        ...
        new com.mongodb.BasicDBObject($MAP);
```

Reply:
Yes that was the playground reformatting, my original submission used your recommended style: https://semgrep.dev/playground/s/aabashkin:mongodb-nosqli

Reply:
Fixed in 67842403b5064740657a5810ea9576bc15f1f84c